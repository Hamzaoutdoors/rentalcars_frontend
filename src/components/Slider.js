import { useEffect } from 'react';
import styled from 'styled-components';
import { ArrowLeftOutlined, ArrowRightOutlined } from '@material-ui/icons';
import { useSelector } from 'react-redux';
import CarCard from './cars/CarCard';
import { mobile } from '../responsive';
import store from '../redux/configureStore';
import { setItemSize, setSliderIndex } from '../redux/utils/actions/sliderActions';

const Container = styled.div`
    display: flex;
    justify-content: center;
    align-items: center;
    position: relative;
    height: 60%;
    width: 100%;
    overflow-x: hidden;
`;

const Arrow = styled.div`
    position: absolute;
    pointer-events: ${(props) => (((props.slideIndex === 0 && props.direction === 'left')
     || ((props.slideIndex === props.carsLength - props.size) && props.direction === 'right')) && 'none')};
    width: 80px;
    height: 50px;
    border-top-${(props) => (props.direction === 'left' ? 'right' : 'left')}-radius: 180px;
    border-bottom-${(props) => (props.direction === 'left' ? 'right' : 'left')}-radius: 180px;
    padding: ${(props) => (props.direction === 'left' ? '0 0 0 20px' : '0 20px 0 0')};
    background-color: #97BF11;
    display: flex;
    justify-content: center;
    align-items: center;
    top: -50px;
    bottom: 0;
    left: ${(props) => (props.direction === 'left' && '10px')};
    right: ${(props) => (props.direction === 'right' && '10px')};
    margin: auto;
    cursor: pointer;
    opacity: 0.5;
    z-index: 2;
    transition: all 0.5s ease-in-out;
    &:hover {
      transform: translateX(${(props) => (props.direction === 'left' ? '-3px' : '3px')});
        opacity: 1;
    }
    ${mobile({
    width: '50px',
  })};
`;

const Wrapper = styled.div`
    display: flex;
    justify-content: space-around;
    align-items: center;
    padding: 0 5%;
    width: 100%;
    transition: all 1.5s ease-in-out;
`;

const dispatchItemSize = (action, size) => {
  store.dispatch(action(size));
};

const dispatchSliderIndex = (action, index) => {
  store.dispatch(action(index));
};

const Slider = () => {
  const { slideIndex, itemsSize } = useSelector((state) => state.utils.slider);
  const cars = useSelector((state) => state.cars.data);

  const handleResize = () => {
    if (window.innerWidth > 1350) {
      dispatchItemSize(setItemSize, 3);
    } else if (window.innerWidth > 700) {
      dispatchItemSize(setItemSize, 2);
    } else {
      dispatchItemSize(setItemSize, 1);
    }
  };

  useEffect(() => {
    handleResize();
    window.addEventListener('resize', () => {
      handleResize();
    });
  }, []);

  const handleNext = (direction) => {
    if (direction === 'left' || slideIndex === cars.length - itemsSize) {
      dispatchSliderIndex(setSliderIndex, slideIndex - 1);
    } else {
      dispatchSliderIndex(setSliderIndex, slideIndex + 1);
    }
  };

  return (
    <Container>
      <Arrow direction="left" size={itemsSize} onClick={() => handleNext('left')} slideIndex={slideIndex} carsLength={cars.length}>
        <ArrowLeftOutlined style={{ fontSize: '3rem', color: '#e6e6e6' }} />
      </Arrow>
      <Wrapper slideIndex={slideIndex}>
        {cars.slice(slideIndex, slideIndex + itemsSize).map((item) => (
          <CarCard key={item.id} item={item} />))}
      </Wrapper>
      <Arrow direction="right" size={itemsSize} onClick={() => handleNext('right')} slideIndex={slideIndex} carsLength={cars.length}>
        <ArrowRightOutlined style={{ fontSize: '3rem', color: '#e6e6e6' }} />
      </Arrow>
    </Container>
  );
};

export default Slider;
