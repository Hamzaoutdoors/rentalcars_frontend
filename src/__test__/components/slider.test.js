import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import Slider from '../../components/Slider';
import store from '../../redux/configureStore';

it('Slider component renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <Slider />
        </Router>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
