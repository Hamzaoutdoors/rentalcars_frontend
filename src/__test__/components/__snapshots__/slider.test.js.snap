// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider component renders correctly 1`] = `
<div
  className="sc-gKclnd hCiNti"
>
  <div
    className="sc-iCfMLu Xfhrl"
    direction="left"
    onClick={[Function]}
    size={3}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      style={
        Object {
          "color": "#e6e6e6",
          "fontSize": "3rem",
        }
      }
      viewBox="0 0 24 24"
    >
      <path
        d="M14 7l-5 5 5 5V7z"
      />
    </svg>
  </div>
  <div
    className="sc-furwcr kpeYjR"
  />
  <div
    className="sc-iCfMLu glyfKO"
    direction="right"
    onClick={[Function]}
    size={3}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      style={
        Object {
          "color": "#e6e6e6",
          "fontSize": "3rem",
        }
      }
      viewBox="0 0 24 24"
    >
      <path
        d="M10 17l5-5-5-5v10z"
      />
    </svg>
  </div>
</div>
`;
